### PR TITLE
net annotations-gen: Cleanup NIC hotplug / unplug logic

### PR DESCRIPTION
**Pull request by orelmisan**
### What this PR does
PR #13235 had moved the NIC hotplug / unplug logic under the network annotations generator.
- Move `CalculateInterfacesAndNetworksForMultusAnnotationUpdate` from `pkg/network/vmispec` closer to its only caller.
- Drop `CalculateInterfacesAndNetworksForMultusAnnotationUpdate`'s unit tests in favor of testing it in the context of its caller.
- Rename `CalculateInterfacesAndNetworksForMultusAnnotationUpdate`.

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

**Comment by orelmisan** [comment]
[Change](https://github.com/kubevirt/kubevirt/compare/bd380b7516dff554ae18545bbc63ffd3884e8a57..8b1dee23d64412eb4ba9d865a65634782c19be04): Adjusted variable name per @EdDev's [suggestion](https://github.com/kubevirt/kubevirt/pull/13235#discussion_r1841817321).

File: pkg/network/vmispec/hotplug_test.go [hunk 1]
```diff
@@ -80,135 +75,4 @@ var _ = Describe("utilitary funcs to identify attachments to hotplug", func() {
 			),
 		)
 	})
-
-	Context("CalculateInterfacesAndNetworksForMultusAnnotationUpdate", func() {
```
[the remark this answers is not in the document]
Comment (orelmisan) [reply pkg/network/vmispec/hotplug_test.go:84]: Following this move, it is tested in the context of `Generator.GenerateFromActivePod`.

The commit message states:
> Move it closer to its usage and test it as part of its caller.

**Comment by orelmisan** [comment]
[Change](https://github.com/kubevirt/kubevirt/compare/8b1dee23d64412eb4ba9d865a65634782c19be04..e85729e1aff3aa51a23915fbb48204ec45023f8b): Moved misplaced include statement.

File: pkg/network/vmispec/hotplug_test.go [hunk 2]
```diff
@@ -80,135 +75,4 @@ var _ = Describe("utilitary funcs to identify attachments to hotplug", func() {
...
-						{"interface":"net1", "name":"red-net", "namespace": "default"},
-						{"interface":"net2", "name":"blue-net", "namespace": "default"}
-					]`,
-					}},
-				},
-				[]v1.Interface{{Name: testNetworkName1, InterfaceBindingMethod: v1.InterfaceBindingMethod{SRIOV: &v1.InterfaceSRIOV{}}}, {Name: testNetworkName2}, {Name: testNetworkName3}},
-				[]v1.Network{{Name: testNetworkName1}, {Name: testNetworkName2}, {Name: testNetworkName3}},
-				expectToChange,
-			),
-			Entry("when vmi interfaces have an extra SRIOV interface which requires hotplug, change is not required since SRIOV hotplug to a pod is not supported",
```
Comment (EdDev) [inline pkg/network/vmispec/hotplug_test.go:141]: I think you added explicitly this scenario in the prev commit to the caller function on which you depend now. But you have reduced it a bit, why? In retro, this move/conversion should have been done in one commit to emphasize what moved (e.g. in the prev commit, to delete this test from here).

Anyway, as a reviewer, I am suppose to check in depth that all the other tests here, with all their details are covered at the caller. If you already done that, please provide the relevant ref to help me as well.

If it was not done, then I would expect all these cases to be exercised on the caller if possible. If not possible, then mention which ones and why.

**Comment by orelmisan** [comment]
[Change](https://github.com/kubevirt/kubevirt/compare/e85729e1aff3aa51a23915fbb48204ec45023f8b..9b436ed4c76edd58d10fadd550da4fc4d68a565c): Rebase

**Comment by orelmisan** [comment]
[Change](https://github.com/kubevirt/kubevirt/compare/9b436ed4c76edd58d10fadd550da4fc4d68a565c..32ed1ce19af69c950d11d367fa5752a5292fe19e): Added two missing test cases.

File: pkg/network/vmispec/hotplug_test.go (hunk 2 above)
Comment (orelmisan) [reply pkg/network/vmispec/hotplug_test.go:141]: >  In retro, this move/conversion should have been done in one commit to emphasize what moved (e.g. in the prev commit, to delete this test from here).

Done.
Same for the other tests.

> Anyway, as a reviewer, I am suppose to check in depth that all the other tests here, with all their details are covered at the caller. If you already done that, please provide the relevant ref to help me as well.

If it was not done, then I would expect all these cases to be exercised on the caller if possible. If not possible, then mention which ones and why.

Done.

File: pkg/network/pod/annotations/generator_test.go [hunk 3]
```diff
@@ -520,6 +520,22 @@ var _ = Describe("Annotations Generator", func() {
 			clusterConfig = newClusterConfig(kv)
 		})
 
+		It("Should not generate network attachment annotation when there are no networks", func() {
+			vmi := libvmi.New(
+				libvmi.WithNamespace(testNamespace),
+				libvmi.WithAutoAttachPodInterface(false),
```
Comment (EdDev) [inline pkg/network/pod/annotations/generator_test.go:528]: How is this relevant? I think this has influence only when the VM controller computes the VMI spec.
It was also not there before and I think the implementation its is not effected by this knob.

**Review by EdDev — COMMENTED** [review]
Nice, thank you!

There is one small point I commented on, but it is probably not worth holding this PR over. Still, try to have a look.

/approve

File: pkg/network/pod/annotations/generator_test.go (hunk 3 above)
Comment (orelmisan) [reply pkg/network/pod/annotations/generator_test.go:528]: This configuration option is required to get to this scenario in a running system and was added for completeness.
It does not affect this logic, so if you think it adds noise I can remove it.

Comment (EdDev) [reply pkg/network/pod/annotations/generator_test.go:528]: On a running system, that is correct for a VM but not for a VMI. I’m in favor of completeness but in this case I find it confusing.

Comment (orelmisan) [reply pkg/network/pod/annotations/generator_test.go:528]: > On a running system, that is correct for a VM but not for a VMI.

The VMI mutator uses it when creating a VMI on a live system.

Comment (orelmisan) [reply pkg/network/pod/annotations/generator_test.go:528]: https://github.com/kubevirt/kubevirt/blob/a724aff0139167ee3997d55ee90cbbb4439aafcd/pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator.go#L85

https://github.com/kubevirt/kubevirt/blob/a724aff0139167ee3997d55ee90cbbb4439aafcd/pkg/defaults/defaults.go#L111

https://github.com/kubevirt/kubevirt/blob/a724aff0139167ee3997d55ee90cbbb4439aafcd/pkg/defaults/defaults.go#L232

Comment (EdDev) [reply pkg/network/pod/annotations/generator_test.go:528]: Thank you for correcting me. I am surprised again and again about how nasty reality is.

I would like to drop this behavior from the VMI.
